src/App.jsx

```diff
@@ -7,22 +7,48 @@ HeadingH1.propTypes = {
   children: PropTypes.node.isRequired,
 };
 
-function useSemiPersistentState() {
-  const [todoList, setToDoList] = useState(() => {
-    const savedList = localStorage.getItem("savedTodoList");
+export default function App() {
+  const [todoList, setToDoList] = useState([]);
+  const [isLoading, setIsLoading] = useState(true);
+
+  // Mimicking a side-effect operation with a delay
+  function sideEffectHandler() {
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        const success = true;
 
-    return savedList ? JSON.parse(savedList) : [];
-  });
+        if (success) {
+          resolve();
+        } else {
+          reject("An error occurred during the operation.");
+        }
+      }, 2000);
+    });
+  }
 
   useEffect(() => {
-    localStorage.setItem("savedTodoList", JSON.stringify(todoList));
-  }, [todoList]);
+    const savedTodos = JSON.parse(localStorage.getItem("todoList"));
+    if (savedTodos) {
+      setToDoList(savedTodos);
+      setIsLoading(false);
+    }
```

Comment:
Isn't it removing the loading state before the 2 seconds of the `sideEffectHandler` call?

I believe you don't need this part of the code. The state will be assigned when the `sideEffectHandler` call resolve.

Reply:
Thanks, I got it now.